File: 0c-bason/assignment-1/cpp/src/bason_codec.cpp
#include "bason_codec.h"

#include <stdexcept>
#include <cstring>

////////////////////////////////////////////////////////////////////////////////

namespace NBason {

////////////////////////////////////////////////////////////////////////////////

namespace {

// BASON tag bytes
constexpr uint8_t TAG_BOOLEAN_SHORT = 'b';  // 0x62
constexpr uint8_t TAG_ARRAY_SHORT   = 'a';  // 0x61
constexpr uint8_t TAG_STRING_SHORT  = 's';  // 0x73
constexpr uint8_t TAG_OBJECT_SHORT  = 'o';  // 0x6F
constexpr uint8_t TAG_NUMBER_SHORT  = 'n';  // 0x6E

constexpr uint8_t TAG_BOOLEAN_LONG  = 'B';  // 0x42
constexpr uint8_t TAG_ARRAY_LONG    = 'A';  // 0x41
constexpr uint8_t TAG_STRING_LONG   = 'S';  // 0x53
constexpr uint8_t TAG_OBJECT_LONG   = 'O';  // 0x4F
constexpr uint8_t TAG_NUMBER_LONG   = 'N';  // 0x4E

constexpr size_t MAX_SHORT_LENGTH = 15;
constexpr size_t MAX_LONG_KEY_LENGTH = 255;

////////////////////////////////////////////////////////////////////////////////

uint8_t GetTagForType(EBasonType type, bool shortForm)
{
    if (shortForm) {
        switch (type) {
            case EBasonType::Boolean: return TAG_BOOLEAN_SHORT;
            case EBasonType::Array:   return TAG_ARRAY_SHORT;
            case EBasonType::String:  return TAG_STRING_SHORT;
            case EBasonType::Object:  return TAG_OBJECT_SHORT;
            case EBasonType::Number:  return TAG_NUMBER_SHORT;
        }
    } else {
        switch (type) {
            case EBasonType::Boolean: return TAG_BOOLEAN_LONG;
            case EBasonType::Array:   return TAG_ARRAY_LONG;
            case EBasonType::String:  return TAG_STRING_LONG;
            case EBasonType::Object:  return TAG_OBJECT_LONG;
            case EBasonType::Number:  return TAG_NUMBER_LONG;
        }
    }
    throw std::logic_error("Unknown BASON type");
}

EBasonType GetTypeFromTag(uint8_t tag)
{
    switch (tag) {
        case TAG_BOOLEAN_SHORT:
        case TAG_BOOLEAN_LONG:
            return EBasonType::Boolean;
        case TAG_ARRAY_SHORT:
        case TAG_ARRAY_LONG:
            return EBasonType::Array;
        case TAG_STRING_SHORT:
        case TAG_STRING_LONG:
            return EBasonType::String;
        case TAG_OBJECT_SHORT:
        case TAG_OBJECT_LONG:
            return EBasonType::Object;
        case TAG_NUMBER_SHORT:
        case TAG_NUMBER_LONG:
            return EBasonType::Number;
        default:
            throw std::runtime_error("Invalid BASON tag: 0x" + std::to_string(tag));
    }
}

bool IsShortForm(uint8_t tag)
{
    return tag >= 'a' && tag <= 'z';
}

bool IsContainerType(EBasonType type)
{
    return type == EBasonType::Array || type == EBasonType::Object;
}

void WriteLittleEndian32(std::vector<uint8_t>& buffer, uint32_t value)
{
    buffer.push_back(value & 0xFF);
    buffer.push_back((value >> 8) & 0xFF);
    buffer.push_back((value >> 16) & 0xFF);
    buffer.push_back((value >> 24) & 0xFF);
}

uint32_t ReadLittleEndian32(const uint8_t* data)
{
    return static_cast<uint32_t>(data[0]) |
           (static_cast<uint32_t>(data[1]) << 8) |
           (static_cast<uint32_t>(data[2]) << 16) |
           (static_cast<uint32_t>(data[3]) << 24);
}

std::vector<uint8_t> EncodeChildren(const std::vector<TBasonRecord>& children)
{
    std::vector<uint8_t> result;
    for (const auto& child : children) {
        auto encoded = EncodeBason(child);
        result.insert(result.end(), encoded.begin(), encoded.end());
    }
    return result;
}

} // namespace

////////////////////////////////////////////////////////////////////////////////

std::vector<uint8_t> EncodeBason(const TBasonRecord& record)
{
    std::vector<uint8_t> result;

    // For containers, encode children first to know the value length
    std::vector<uint8_t> valueData;
    
    if (IsContainerType(record.Type)) {
        valueData = EncodeChildren(record.Children);
    } else {
        valueData.assign(record.Value.begin(), record.Value.end());
    }

    size_t keyLen = record.Key.size();
    size_t valueLen = valueData.size();

    // Choose short or long form
    bool useShortForm = (keyLen <= MAX_SHORT_LENGTH && valueLen <= MAX_SHORT_LENGTH);

    // Write tag
    uint8_t tag = GetTagForType(record.Type, useShortForm);
    result.push_back(tag);

    if (useShortForm) {
        // Short form: lengths byte (upper nibble = key len, lower nibble = value len)
        uint8_t lengths = (keyLen << 4) | valueLen;
        result.push_back(lengths);
    } else {
        // Long form: val_len (4B LE) + key_len (1B)
        if (keyLen > MAX_LONG_KEY_LENGTH) {
            throw std::invalid_argument("Key length exceeds maximum (255 bytes)");
        }
        if (valueLen > UINT32_MAX) {
            throw std::invalid_argument("Value length exceeds maximum (4GB)");
        }
        
        WriteLittleEndian32(result, static_cast<uint32_t>(valueLen));
        result.push_back(static_cast<uint8_t>(keyLen));
    }

    // Write key
    result.insert(result.end(), record.Key.begin(), record.Key.end());

    // Write value
    result.insert(result.end(), valueData.begin(), valueData.end());

    return result;
}

std::pair<TBasonRecord, size_t> DecodeBason(const uint8_t* data, size_t len)
{
    if (len < 2) {
        throw std::runtime_error("Insufficient data for BASON record (need at least 2 bytes)");
    }

    TBasonRecord record;
    size_t offset = 0;

    // Read tag
    uint8_t tag = data[offset++];
    record.Type = GetTypeFromTag(tag);
    bool shortForm = IsShortForm(tag);

    size_t keyLen, valueLen;

    if (shortForm) {
        // Short form
        uint8_t lengths = data[offset++];
        keyLen = (lengths >> 4) & 0x0F;
        valueLen = lengths & 0x0F;
    } else {
        // Long form
        if (len < 6) {
            throw std::runtime_error("Insufficient data for long form BASON record");
        }
        valueLen = ReadLittleEndian32(&data[offset]);
        offset += 4;
        keyLen = data[offset++];
    }

    // Read key
    if (offset + keyLen > len) {
        throw std::runtime_error("Insufficient data for key");
    }
    record.Key.assign(reinterpret_cast<const char*>(&data[offset]), keyLen);
    offset += keyLen;

    // Read value
    if (offset + valueLen > len) {
        throw std::runtime_error("Insufficient data for value");
    }

    if (IsContainerType(record.Type)) {
        // Decode children
        size_t childOffset = 0;
        while (childOffset < valueLen) {
            auto [child, childSize] = DecodeBason(&data[offset + childOffset], valueLen - childOffset);
            record.Children.push_back(std::move(child));
            childOffset += childSize;
        }
    } else {
        // Leaf value
        record.Value.assign(reinterpret_cast<const char*>(&data[offset]), valueLen);
    }
    offset += valueLen;

    return {std::move(record), offset};
}

std::vector<TBasonRecord> DecodeBasonAll(const uint8_t* data, size_t len)
{
    std::vector<TBasonRecord> records;
    size_t offset = 0;

    while (offset < len) {
        auto [record, size] = DecodeBason(&data[offset], len - offset);
        records.push_back(std::move(record));
        offset += size;
    }

    return records;
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NBason

File: 0c-bason/assignment-1/go/pkg/bason/bason_codec.go
package bason

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// BasonType represents the BASON value type
type BasonType int

const (
	TypeBoolean BasonType = iota
	TypeArray
	TypeString
	TypeObject
	TypeNumber
)

// Record is a single BASON record (TLV with key and value or children)
type Record struct {
	Type     BasonType
	Key      string
	Value    string // For leaf types
	Children []Record
}

// BASON tag bytes
const (
	tagBooleanShort = 'b'
	tagArrayShort   = 'a'
	tagStringShort  = 's'
	tagObjectShort  = 'o'
	tagNumberShort  = 'n'

	tagBooleanLong = 'B'
	tagArrayLong   = 'A'
	tagStringLong  = 'S'
	tagObjectLong  = 'O'
	tagNumberLong  = 'N'

	maxShortLength   = 15
	maxLongKeyLength = 255
)

var (
	ErrInsufficientData = errors.New("insufficient data for BASON record")
	ErrInvalidTag       = errors.New("invalid BASON tag")
	ErrKeyTooLong       = errors.New("key length exceeds maximum (255 bytes)")
)

func getTagForType(t BasonType, shortForm bool) byte {
	if shortForm {
		switch t {
		case TypeBoolean:
			return tagBooleanShort
		case TypeArray:
			return tagArrayShort
		case TypeString:
			return tagStringShort
		case TypeObject:
			return tagObjectShort
		case TypeNumber:
			return tagNumberShort
		}
	} else {
		switch t {
		case TypeBoolean:
			return tagBooleanLong
		case TypeArray:
			return tagArrayLong
		case TypeString:
			return tagStringLong
		case TypeObject:
			return tagObjectLong
		case TypeNumber:
			return tagNumberLong
		}
	}
	panic("unknown BASON type")
}

func getTypeFromTag(tag byte) (BasonType, error) {
	switch tag {
	case tagBooleanShort, tagBooleanLong:
		return TypeBoolean, nil
	case tagArrayShort, tagArrayLong:
		return TypeArray, nil
	case tagStringShort, tagStringLong:
		return TypeString, nil
	case tagObjectShort, tagObjectLong:
		return TypeObject, nil
	case tagNumberShort, tagNumberLong:
		return TypeNumber, nil
	default:
		return 0, fmt.Errorf("%w: 0x%02x", ErrInvalidTag, tag)
	}
}

func isShortForm(tag byte) bool {
	return tag >= 'a' && tag <= 'z'
}

func isContainerType(t BasonType) bool {
	return t == TypeArray || t == TypeObject
}

func readLittleEndian32(data []byte) uint32 {
	return binary.LittleEndian.Uint32(data)
}

func encodeChildren(children []Record) ([]byte, error) {
	var result []byte
	for _, child := range children {
		enc, err := EncodeBason(child)
		if err != nil {
			return nil, err
		}
		result = append(result, enc...)
	}
	return result, nil
}

// EncodeBason encodes a record to bytes.
// The encoder uses short form when both key and value fit in 15 bytes
func EncodeBason(record Record) ([]byte, error) {
	var valueData []byte
	var err error
	if isContainerType(record.Type) {
		valueData, err = encodeChildren(record.Children)
		if err != nil {
			return nil, err
		}
	} else {
		valueData = []byte(record.Value)
	}

	keyLen := len(record.Key)
	valueLen := len(valueData)

	useShortForm := keyLen <= maxShortLength && valueLen <= maxShortLength

	var result []byte
	tag := getTagForType(record.Type, useShortForm)
	result = append(result, tag)

	if useShortForm {
		lengths := byte(keyLen<<4) | byte(valueLen)
		result = append(result, lengths)
	} else {
		if keyLen > maxLongKeyLength {
			return nil, ErrKeyTooLong
		}
		valLenBuf := make([]byte, 4)
		binary.LittleEndian.PutUint32(valLenBuf, uint32(valueLen))
		result = append(result, valLenBuf...)
		result = append(result, byte(keyLen))
	}

	result = append(result, []byte(record.Key)...)
	result = append(result, valueData...)
	return result, nil
}

// DecodeBason decodes a single record from data.
// Returns the record and the number of bytes consumed
func DecodeBason(data []byte) (Record, int, error) {
	if len(data) < 2 {
		return Record{}, 0, fmt.Errorf("%w (need at least 2 bytes)", ErrInsufficientData)
	}

	offset := 0
	tag := data[offset]
	offset++

	recordType, err := getTypeFromTag(tag)
	if err != nil {
		return Record{}, 0, err
	}
	shortForm := isShortForm(tag)

	var keyLen, valueLen int
	if shortForm {
		lengths := data[offset]
		offset++
		keyLen = int(lengths>>4) & 0x0F
		valueLen = int(lengths & 0x0F)
	} else {
		if len(data) < 6 {
			return Record{}, 0, fmt.Errorf("%w for long form", ErrInsufficientData)
		}
		valueLen = int(readLittleEndian32(data[offset:]))
		offset += 4
		keyLen = int(data[offset])
		offset++
	}

	if offset+keyLen > len(data) {
		return Record{}, 0, fmt.Errorf("%w for key", ErrInsufficientData)
	}
	record := Record{
		Type: recordType,
		Key:  string(data[offset : offset+keyLen]),
	}
	offset += keyLen

	if offset+valueLen > len(data) {
		return Record{}, 0, fmt.Errorf("%w for value", ErrInsufficientData)
	}

	if isContainerType(record.Type) {
		childOffset := 0
		for childOffset < valueLen {
			child, childSize, err := DecodeBason(data[offset+childOffset : offset+valueLen])
			if err != nil {
				return Record{}, 0, err
			}
			record.Children = append(record.Children, child)
			childOffset += childSize
		}
	} else {
		record.Value = string(data[offset : offset+valueLen])
	}
	offset += valueLen

	return record, offset, nil
}

// DecodeBasonAll decodes all records from a byte buffer (concatenated records)
func DecodeBasonAll(data []byte) ([]Record, error) {
	var records []Record
	offset := 0
	for offset < len(data) {
		record, size, err := DecodeBason(data[offset:])
		if err != nil {
			return nil, err
		}
		records = append(records, record)
		offset += size
	}
	return records, nil
}
